### ISSUE TITLE
hierarchical editor operations performance

### ISSUE THREAD

**Issue by dustymc** [ArctosDB/arctos#3607]
ref: https://github.com/ArctosDB/arctos/issues/3538#issuecomment-843349240

The hierarchical editor is relatively slow with delete and import. 15K terms work, slowly, much larger probably won't.

Large operations need made asynchronous, probably via a new task and controller.

Workaround: send me data, I'll figure it out.

Alternate workaround: Load smaller files, but this introduces some tricky dependencies due to parentage.

Needs prioritized, tentatively suggest waiting on an actual need; real data would be extremely useful for this.

**Comment by dustymc** [comment]
Very tentatively recommending we dontfix this and deprecate the hierarchical editor.

Why this is a horrible idea:

* The hierarchical editor is still the only structurally consistent option for managing classifications.

Why that doesn't matter:

* The classification loader is very good; it will handle very wide (20 + 60, easily extendable) classifications of virtually any length - certainly a few million rows would be no challenge.
* We no longer have strict vocabulary controls - there's no need to figure out how we would spell something, just enter what needs to be there.
* Given one or a few users managing relatively small batches of relatively homogeneous data in some spreadsheet-like application (Google Sheets is probably more secure than your hard drive and makes collaboration easy), it should be relatively straightforward to maintain consistency even without controls.
* Managing whole collection Sources is no longer required.
* Given those smaller classifications, whatever structural control the hierarchical editor provides is perhaps less important than being able to see everything in a familiar format.
* We have lots of sources of data, users will still (sometimes, probably) find stuff even when its managed under an inconsistent classification.

In other words, I think managing classifications as smaller more homogeneous batches (perhaps insects or molluscs, just not mixed with plants or fish) and combining those into whatever is required for cataloging is just _better_, and that includes not requiring sophisticated/dedicated tools.

**Comment by Jegelewicz** [comment]
We need this - maybe we should look at the structure and presentation used by Symbiota.

Watch - https://youtu.be/ipSIXOHud8A?t=534

Code - https://github.com/BioKIC/Symbiota/tree/master/taxa

This seems to work really well, but the difficulty them comes in getting from the tree to Arctos taxonomy format.

I'd recommend watching this whole thing because I kinda like the way they are doing this from a discovery standpoint and allowing people to ID stuff as whatever the hell they want, then attempt to match up with the thesaurus. Note that they also have issues with homonyms. Missing is the ability to add a dwc:scientificNameID.

**Comment by Jegelewicz** [comment]
An honest question from @ArctosDB/taxonomy today.

Why have structured taxonomic classifications as separate things rather than the tree structure that literally the rest of the world uses?

**Comment by dustymc** [comment]
Taxonomy at scale is not hierarchical. There's a very lengthy discussion somewhere, maybe in the google group.

**Comment by Jegelewicz** [comment]
I think we are the only people in the world who believe this.

**Comment by Jegelewicz** [comment]
And I am not saying that I do.

**Comment by dustymc** [comment]
>only people in the world who believe this.

It's not a faith-based system....

**Comment by Jegelewicz** [comment]
@wellerjes @sharpphyl @adhornsby @genevieve-anderegg feel free to weigh in!

**Comment by wellerjes** [comment]
From my understanding, if we have a tree structure, then it eliminates problems like this:
![image](https://user-images.githubusercontent.com/93737929/189432362-5363a44d-8dea-4388-afbd-fd5ce1960875.png)

where the taxonomy is incomplete and needs to be updated manually (refresh and pull global names). It's going to find and update each taxon record--and even then, there is a higher risk of misspellings/errors because it's been done manually. I'd rather have a set taxonomy structure that we can clone for the one-offs where taxonomists don't agree, over everything being open.

**Comment by acdoll** [comment]
> help you get started with a new Source

So we can develop our own taxonomic structure in a google sheet (eg "DMNS Taxonomy") and Arctos can pull from that? And add/edit that sheet at-will and Arctos will reincorporate (maybe with a manual refresh?)?

**Comment by Jegelewicz** [comment]
> Arctos can pull from that?

This could possibly be a thing, but I do know you can import at any time and overwrite whatever you want. BUT WHY? We are slowly devolving into everyone just managing their own taxonomy and losing the community effort.

Also - you will have to either keep it in Arctos preferred format or have a process for taking whatever you keep into the 
[classification bulkload format](https://arctos.database.museum/tools/BulkloadClassification.cfm).

This is the challenge I faced working on the TPT taxonomy stuff - they maintain a tree structure and going from that to the Arctos create a full classification for every name structure is not always as simple as it might seem - then there is the classification metadata....

.

**Comment by acdoll** [comment]
I'm not saying _I_ want to do that, but just trying to be clear on what is possible or what Dusty was suggesting.

**Comment by dustymc** [comment]
Yes.

Set up a Google Sheet, periodically upload to Arctos, prefer it before other Sources that might contain those data.

You can directly share by giving any level of access to whomever you want, and they probably won't need any training because it's just a spreadsheet.

Collections who don't directly participate can still share by using your Source, and hopefully they'll develop something for some other taxa/group/whatever that you can use.

My hope is that this will result in collections (or individuals at collections) curating what they care about without getting completely overwhelmed by thousands of things they don't really care about. If you're working on North American shrews (or WHATEVER) you can just curate that without getting tangled up in giraffes and bats and whatever else - and again hopefully some other collection will pick those up if this catches on.

Things like AOU's checklist fit nicely in there too (and you can even prefer your '6 birds that I think AOU got wrong' source before it, if you don't like the whole thing).

I'm happy to help with mapping, either as a one-time thing or potentially as Arctos tools.

**Comment by Jegelewicz** [comment]
@dustymc is there any hope of the hierarchical editor functioning? If not, I suggest we remove the tool and close this.

**Comment by dustymc** [comment]
It works (AFAIK), but it also looks like an unnecessary complication to me at this point. I'd be in favor of deprecating it, unless there's some compelling use case. (And it's been central to leading us to a place from which is doesn't look so critical, I certainly don't see anything about it as wasted effort!)

**Comment by Jegelewicz** [comment]
HMMM - OK I can see how it is a bit indecipherable though. I wonder if we could take just one group that @DerekSikes is interested in (Carabidae), get a download from Arctos and import it to the tool so we can demonstrate it's potential usefulness?

Get me download from Arctos for that group and I'll see if I can do that?

**Comment by dustymc** [comment]
The tool clearly isn't being used and is greatly complicating things like https://github.com/ArctosDB/arctos/issues/6302, I'm proceeding unless there's some compelling nontheoretical reason to retain this.
